File: agent/assembler.py
def assemble_brief(
    date_str: str,
    parsed_tasks: dict,
    parsed_goals: dict,
    all_tactical_rules: list,
    all_emotional_rules: list,
    all_hobby_tasks: list,
    all_technical_tasks: list,
    yesterday_wins: list,
    llm_output: dict | None,
    config: dict,
) -> dict:
    fields = config["fields"]

    if fields["tactical_rules"]["mode"] == "llm" and llm_output:
        tactical_rules = llm_output["tactical_rules"]
    else:
        count = fields["tactical_rules"].get("count", 4)
        tactical_rules = all_tactical_rules[:count]

    if fields["identity_rules"]["mode"] == "llm" and llm_output:
        identity_rules = llm_output["identity_rules"]
    else:
        count = fields["identity_rules"].get("count", 3)
        identity_rules = all_emotional_rules[:count]

    if fields["goals"]["mode"] == "llm" and llm_output:
        goals = llm_output["goals"]
    else:
        goals = parsed_goals

    if fields["hobby_tasks"]["mode"] == "llm" and llm_output:
        hobby_tasks = llm_output["hobby_tasks"]
    else:
        count = fields["hobby_tasks"].get("count", 3)
        hobby_tasks = all_hobby_tasks[:count]

    if fields["suggestion"]["mode"] == "llm" and llm_output:
        suggestion = llm_output.get("suggestions", [])
    else:
        suggestion = []

    wins = yesterday_wins

    return {
        "date": date_str,
        "identity": {"rules": identity_rules},
        "goals": goals,
        "tasks": parsed_tasks,
        "tactical_rules": tactical_rules,
        "technical_tasks": all_technical_tasks[:config["fields"].get("technical_tasks", {}).get("count", 5)],
        "hobby_tasks": hobby_tasks,
        "suggestions": suggestion,
        "wins": wins,
    }

File: src/agent/assembler.ts
import type { MorningOSSettings } from "../settings";
import type { DailyBrief, SuggestionSource } from "../types";
import type { ParsedGoals } from "./vault-reader";

const VALID_SOURCES = new Set<string>(["tasks", "goals", "technical_backlog", "wins"]);

export interface LLMOutput {
  tactical_rules?: string[];
  identity_rules?: string[];
  suggestions?: { text: string; source: string }[];
  goals?: { short_term?: string[]; long_term?: string[] };
  wins?: string[];
}

export function assembleBrief(
  dateStr: string,
  parsedGoals: ParsedGoals,
  allTacticalRules: string[],
  allEmotionalRules: string[],
  yesterdayWins: string[],
  llmOutput: LLMOutput | null,
  settings: MorningOSSettings
): DailyBrief {
  const tacticalRules =
    settings.modeTacticalRules && llmOutput?.tactical_rules
      ? llmOutput.tactical_rules
      : allTacticalRules.slice(0, settings.tacticalRulesCount);

  const identityRules =
    settings.modeIdentityRules && llmOutput?.identity_rules
      ? llmOutput.identity_rules
      : allEmotionalRules.slice(0, settings.identityRulesCount);

  const goals =
    settings.modeGoals && llmOutput?.goals
      ? {
          short_term: llmOutput.goals.short_term ?? parsedGoals.short_term,
          long_term: llmOutput.goals.long_term ?? parsedGoals.long_term,
        }
      : parsedGoals;

  const suggestions =
    settings.modeSuggestion && llmOutput?.suggestions
      ? llmOutput.suggestions.slice(0, settings.suggestionCount).map(s => ({
          text: s.text,
          source: (VALID_SOURCES.has(s.source) ? s.source : "tasks") as SuggestionSource,
        }))
      : [];

  const wins =
    settings.modeWins && llmOutput?.wins
      ? llmOutput.wins
      : yesterdayWins;

  return {
    date: dateStr,
    meta: {
      goals: {
        short_term_count: settings.goalsShortTermCount,
        long_term_count: settings.goalsLongTermCount,
      },
    },
    identity: { rules: identityRules },
    goals,
    tactical_rules: tacticalRules,
    suggestions,
    wins,
  };
}
